src/time_tracker/invoices/handlers/spec.clj

```diff
@@ -1,14 +1,50 @@
 (ns time-tracker.invoices.handlers.spec
   (:require [clojure.spec :as s]
-            [time-tracker.spec :as core-spec]))
-
-(s/def ::start ::core-spec/epoch)
-(s/def ::end ::core-spec/epoch)
-;; A non-empty string.
-(s/def ::client (s/and string?
-                       seq))
-
-(s/def ::generate-invoice-params
-  (s/and (s/keys :req-un [::start ::end ::client])
-         (fn [{:keys [start end]}]
-           (< start end))))
+            [time-tracker.spec :as core-spec]
+            [time-tracker.invoices.spec :as invoices-spec]
+            [time-tracker.users.spec :as users-spec]
+            [time-tracker.timers.spec :as timers-spec]))
+
+(s/def ::rate number?)
+(s/def ::user-id->rate-map-raw (s/keys :req-un [::invoices-spec/user-id ::rate]))
+(s/def ::user-id->rate (s/coll-of ::user-id->rate-map-raw))
+
+(s/def ::tax-percentage number?)
+(s/def ::tax-rate-map-raw (s/keys :req-un [::invoices-spec/tax-name ::tax-percentage]))
+(s/def ::tax-rates (s/nilable (s/coll-of ::tax-rate-map-raw)))
+
+(s/def ::currency string?)
+
+(s/def ::generate-invoice-params-raw
+  (s/merge (s/keys :req-un [::invoices-spec/client
+                            ::invoices-spec/address
+                            ::invoices-spec/notes
+                            ::user-id->rate
+                            ::tax-rates
+                            ::currency
+                            ::core-spec/utc-offset])
+           ::invoices-spec/date-range))
+
+(s/def ::generate-invoice-params-coerced
+  (s/merge (s/keys :req-un [::invoices-spec/client
+                            ::invoices-spec/address
+                            ::invoices-spec/notes
+                            ::invoices-spec/user-id->rate
+                            ::invoices-spec/tax-rates
+                            ::invoices-spec/currency
+                            ::core-spec/utc-offset])
+           ::invoices-spec/date-range))
+
+(defn- invoice-data-pred
+  [{:keys [user-id->rate users] :as invoice-data}]
```

Comment:
oh, sorry, it's _not_ a map. it's a list. even more confusing imho. strive to get to a better name. `->` is an easy cop-out. in this case I read it wrong, even.